File: ig-codegen-python/src/ig_codegen/name_utils.py
"""Converts FHIR resource ids (kebab-case or, occasionally, PascalCase) into Python constant names."""

import re

_LOWER_OR_DIGIT_FOLLOWED_BY_UPPER = re.compile(r"(?<=[a-z0-9])(?=[A-Z])")
_NON_ALPHANUMERIC_RUN = re.compile(r"[^A-Za-z0-9]+")


def to_constant_name(fhir_id: str) -> str:
    with_word_boundaries = _LOWER_OR_DIGIT_FOLLOWED_BY_UPPER.sub("_", fhir_id)
    return _NON_ALPHANUMERIC_RUN.sub("_", with_word_boundaries).upper()


def to_pascal_case(segment: str) -> str:
    """Converts a dot-separated FHIR package name segment into a PascalCase Python class name."""
    with_word_boundaries = _LOWER_OR_DIGIT_FOLLOWED_BY_UPPER.sub("_", segment)
    words = [word for word in _NON_ALPHANUMERIC_RUN.split(with_word_boundaries) if word]
    return "".join(word[0].upper() + word[1:].lower() for word in words)


def to_snake_case(constant_name: str) -> str:
    """Converts a ``SCREAMING_SNAKE_CASE`` constant name (as produced by :func:`to_constant_name`)
    into a lowercase snake_case Python identifier, e.g. ``MII_PR_DIAGNOSE_CONDITION ->
    mii_pr_diagnose_condition``.
    """
    return constant_name.lower()


def to_enum_constant_name(code: str) -> str:
    """Converts a CodeSystem concept code into a valid Python enum member name. Real FHIR code
    systems use codes that aren't valid Python identifiers as-is - e.g. purely numeric codes
    (``"2"``), or receptor-status codes ending in ``+``/``-`` (``"mol+"``, ``"i-"``), which would
    otherwise collide once the sign is stripped by :func:`to_constant_name`.

    This does not guarantee uniqueness across a whole CodeSystem by itself - callers must still
    disambiguate any remaining collisions (e.g. two codes differing only in characters this
    function treats as equivalent).
    """
    if code.endswith("+"):
        base, suffix = code[:-1], "_POS"
    elif code.endswith("-"):
        base, suffix = code[:-1], "_NEG"
    else:
        base, suffix = code, ""

    constant_name = to_constant_name(base) + suffix
    if not constant_name:
        return "_"
    if constant_name[0].isdigit():
        constant_name = f"_{constant_name}"
    return f"{constant_name}_" if _is_sunder(constant_name) else constant_name


def _is_sunder(name: str) -> bool:
    """``enum.Enum`` reserves single-leading/single-trailing-underscore names (``_sunder_``,
    e.g. ``_100_``) for its own use, unlike Java enum constants which have no such restriction.
    A code like ``"100%"`` produces exactly this shape once the leading-digit prefix is added, so
    it needs an extra trailing underscore to stay a plain, unreserved identifier.
    """
    return (
        len(name) > 2
        and name.startswith("_")
        and name.endswith("_")
        and not name.startswith("__")
        and not name.endswith("__")
    )

File: ig-codegen-cs/IgCodegen/NameUtils.cs
using System.Text.RegularExpressions;

namespace IgCodegen;

/// <summary>Converts FHIR resource ids (kebab-case or PascalCase) into C# identifier names.</summary>
public static partial class NameUtils
{
    [GeneratedRegex(@"(?<=[a-z0-9])(?=[A-Z])")]
    private static partial Regex LowerOrDigitFollowedByUpper();

    [GeneratedRegex(@"[^A-Za-z0-9]+")]
    private static partial Regex NonAlphanumericRun();

    [GeneratedRegex(@"_")]
    private static partial Regex Underscore();

    public static string ToConstantName(string id)
    {
        var withWordBoundaries = LowerOrDigitFollowedByUpper().Replace(id, "_");
        return NonAlphanumericRun().Replace(withWordBoundaries, "_").ToUpperInvariant();
    }

    /// <summary>Converts a dot-separated FHIR package name segment into a PascalCase C# identifier.</summary>
    public static string ToPascalCase(string segment)
    {
        var withWordBoundaries = LowerOrDigitFollowedByUpper().Replace(segment, "_");
        var sb = new System.Text.StringBuilder();
        foreach (var word in NonAlphanumericRun().Split(withWordBoundaries))
        {
            if (word.Length == 0)
                continue;
            sb.Append(char.ToUpperInvariant(word[0]));
            if (word.Length > 1)
                sb.Append(word[1..].ToLowerInvariant());
        }
        return sb.ToString();
    }

    /// <summary>
    /// Converts a CodeSystem concept code into a valid C# enum member name, applying the same
    /// <c>+</c>/<c>-</c> suffix disambiguation and digit-prefix guard as the Java counterpart.
    /// </summary>
    public static string ToEnumMemberName(string code)
    {
        string basePart,
            suffix;
        if (code.EndsWith('+'))
        {
            basePart = code[..^1];
            suffix = "_POS";
        }
        else if (code.EndsWith('-'))
        {
            basePart = code[..^1];
            suffix = "_NEG";
        }
        else
        {
            basePart = code;
            suffix = "";
        }

        var name = ToConstantName(basePart) + suffix;
        if (name.Length == 0)
            return "_";
        return char.IsDigit(name[0]) ? "_" + name : name;
    }
}
